src/libs/sys-util-rs/src/cgroup.rs

```diff
@@ -0,0 +1,718 @@
+    ];
+    filter_disabled_cgroup(&mut controllers);
+    Box::new(V1Customized::new(controllers))
+}
+
+/// Get cgroup hierarchy object from `path`.
+pub fn get_hierarchy_by_path(path: &str) -> Result<BoxedHierarchyObject> {
+    let v1 = get_cgroup_hierarchies().clone();
+    let valid_path = valid_cgroup_path(path)?;
+    let cg = cgroups::Cgroup::load(Box::new(v1), valid_path.as_str());
+
+    let mut hierarchy = vec![];
+    for subsys in cg.subsystems() {
+        let controller = subsys.to_controller();
+        if controller.exists() {
+            hierarchy.push(controller.control_type());
+        }
+    }
+
+    Ok(Box::new(V1Customized::new(hierarchy)))
+}
+
+/// Create or load a cgroup object from a path.
+pub fn create_or_load_cgroup(path: &str) -> Result<Cgroup> {
+    let hie = Box::new(get_cgroup_hierarchies().clone());
+
+    create_or_load_cgroup_with_hier(hie, path)
+}
+
+/// Create or load a cgroup v1 object from a path, with a given hierarchy object.
+pub fn create_or_load_cgroup_with_hier(hie: BoxedHierarchyObject, path: &str) -> Result<Cgroup> {
+    let valid_path = valid_cgroup_path(path)?;
+    if is_cgroup_exist(valid_path.as_str()) {
+        Ok(cgroups::Cgroup::load(hie, valid_path.as_str()))
+    } else {
+        Ok(cgroups::Cgroup::new(hie, valid_path.as_str()))
+    }
+}
+
+/// Check whether `path` hosts a cgroup hierarchy directory.
+pub fn is_cgroup_exist(path: &str) -> bool {
+    let valid_path = match valid_cgroup_path(path) {
+        Ok(v) => v,
+        Err(e) => {
+            warn!(sl!(), "{}", e);
+            return false;
+        }
+    };
+
+    let v1 = get_cgroup_hierarchies().clone();
+    let cg = cgroups::Cgroup::load(Box::new(v1), valid_path.as_str());
+    for subsys in cg.subsystems() {
+        if subsys.to_controller().exists() {
+            debug!(sl!(), "cgroup {} exist", path);
+            return true;
+        }
+    }
+
+    false
+}
+
+// Validate the cgroup path is a relative path, do not include ".", "..".
+fn valid_cgroup_path(path: &str) -> Result<String> {
+    let path = path.trim_start_matches('/').to_string();
+
+    for comp in Path::new(&path).components() {
+        if !matches!(comp, Component::Normal(_)) {
+            return Err(Error::InvalidCgroupPath(path.to_string()));
+        }
+    }
+
+    Ok(path)
+}
+
+/// Remove all task from cgroup and delete the cgroup.
+pub fn force_delete_cgroup(cg: cgroups::Cgroup) -> Result<()> {
+    delete_cgroup_with_retry(cg, |cg: &Cgroup| {
+        // if task exist need to delete first.
+        for cg_pid in cg.tasks() {
+            warn!(sl!(), "Delete cgroup task pid {}", cg_pid.pid);
+            cg.remove_task(cg_pid);
+        }
+    })
+}
+
+/// Try to delete a cgroup, call the `do_process` handler at each iteration.
+pub fn delete_cgroup_with_retry<F>(cg: Cgroup, mut do_process: F) -> Result<()>
+where
+    F: FnMut(&Cgroup),
+{
+    // sleep DURATION
+    const DURATION: u64 = 10;
+    const RETRY: u64 = 200;
```

Comment:
Please include the units in these constants (`SLEEP_MILLISECS`, `RETRY_COUNT` or similar).

Reply:
fixed